lib/Dialect/SPIRV/Transforms/DecorateWithLayoutInfoSPIRVPass.cpp

```diff
@@ -0,0 +1,282 @@
+        ptrType.getPointeeType().cast<spirv::StructType>(), baseOffset);
+    auto decoratedType =
+        spirv::PointerType::get(structType, ptrType.getStorageClass());
+    // Save all named attributes except "type" attribute.
+    SmallVector<NamedAttribute, 4> globalVarAttrs;
+
+    for (const auto &attr : op.getAttrs()) {
+      if (attr.first == "type") {
+        continue;
+      }
+      globalVarAttrs.push_back(attr);
+    }
+
+    rewriter.replaceOpWithNewOp<spirv::GlobalVariableOp>(
+        op, rewriter.getTypeAttr(decoratedType), globalVarAttrs);
+    return matchSuccess();
+  }
+};
+
+class SPIRVAddressOfOpLayoutInfoDecoration
+    : public OpRewritePattern<spirv::AddressOfOp> {
+public:
+  using OpRewritePattern<spirv::AddressOfOp>::OpRewritePattern;
+
+  PatternMatchResult matchAndRewrite(spirv::AddressOfOp op,
+                                     PatternRewriter &rewriter) const override {
+    auto spirvModule = op.getParentOfType<spirv::ModuleOp>();
+    if (!spirvModule) {
+      return matchFailure();
+    }
+    auto varName = op.variable();
+    auto varOp = spirvModule.lookupSymbol<spirv::GlobalVariableOp>(varName);
+    if (!varOp) {
+      return matchFailure();
+    }
+
+    rewriter.replaceOpWithNewOp<spirv::AddressOfOp>(
+        op, varOp.type(), rewriter.getSymbolRefAttr(varName));
+    return matchSuccess();
+  }
+};
+} // namespace
+
+void populateSPIRVLayoutInfoPatterns(OwningRewritePatternList &patterns,
+                                     MLIRContext *ctx) {
+  patterns.insert<SPIRVGlobalVariableOpLayoutInfoDecoration,
+                  SPIRVAddressOfOpLayoutInfoDecoration>(ctx);
+}
+
+namespace {
+class DecorateWithLayoutInfoSPIRVPass
+    : public ModulePass<DecorateWithLayoutInfoSPIRVPass> {
+private:
+  void runOnModule() override;
+};
+
+void DecorateWithLayoutInfoSPIRVPass::runOnModule() {
+  auto module = getModule();
+  OwningRewritePatternList patterns;
+  populateSPIRVLayoutInfoPatterns(patterns, module.getContext());
+  ConversionTarget target(*(module.getContext()));
+  target.addLegalDialect<spirv::SPIRVDialect>();
+  target.addLegalOp<FuncOp>();
+  // Type is illegal in terms of legalization if it's a
+  // !spv.ptr<!spv.struct<...>> type in storage class
+  // StorageBuffer/Uniform/PushConstant and does not have layout info.
+  target.addDynamicallyLegalOp<spirv::GlobalVariableOp>(
+      [](spirv::GlobalVariableOp op) {
+        return SPIRVLayoutInfoUtils::isLegalType(op.type());
+      });
+  // Change the type for the direct users.
+  target.addDynamicallyLegalOp<spirv::AddressOfOp>([](spirv::AddressOfOp op) {
+    return SPIRVLayoutInfoUtils::isLegalType(op.pointer()->getType());
+  });
+  // TODO: Change the type for the indirect users such as spv.Load, spv.Store,
+  // spv.FunctionCall and so on.
+  for (auto spirvModule : module.getOps<spirv::ModuleOp>()) {
+    if (failed(applyPartialConversion(spirvModule, target, patterns))) {
```

Comment:
applyPartialConversion is OK, but maybe better to do applyFullConversion. It is better to know that not all global variables conversion was possible.